### DISCUSSION TITLE
Questions on Pdf Generation

### DISCUSSION THREAD

**Discussion by Kazagan** [QuestPDF/QuestPDF#611] [Q&A] [answered]
Starting by attaching Pdf as questions will be based on current state of my PDF.
[PdfExample.pdf](https://github.com/QuestPDF/QuestPDF/files/12097783/PdfExample.pdf)

Brief summary on generation: 
_a 3 Lists of Lists of `Teams`, that is a single list split into 3, to give a flow of top down, left right. 
Row: Loop through each Top list to create the three columns top to bottom.
    Column: Each Row contains columns with items created with it's own loop creating the individual boxes that can be seen 
        within the boxes, with the colors representing data from the objects.        
Each Box of color is defined With column, having the `Team` Name, with a table of the players on the team, and some information about them below the Name._ 

 **Questions start:**
So to ensure each container doesn't break onto the next page, I'm using `ShowEntire()` In the case that a container has too many fields, it will cause the generation to crash. Is there a way to check that it's going to be too large, and break it up into a second container? 

In another example on pages 3 and 4, There is more than enough room for the last element in the second column to fit onto page 3, yet it gets shunted off to the next page. This may be related to `ShowEntire()`, but it was happening before I added that.

**Comment by MarcinZiabek** [comment]
> So to ensure each container doesn't break onto the next page, I'm using ShowEntire() In the case that a container has too many fields, it will cause the generation to crash. Is there a way to check that it's going to be too large, and break it up into a second container?

The `EnsureSpace()` element may help a little. It pushes its child to the next page, if there is less than required vertical space. You may think of it as a less-strict version of the `ShowEntire()` element.

> In another example on pages 3 and 4, There is more than enough room for the last element in the second column to fit onto page 3, yet it gets shunted off to the next page. This may be related to ShowEntire(), but it was happening before I added that.

Indeed, it does not look correct. Would you be able to share a minimal example that reproduces this issue? This may help me finding the problem.

**Comment by Kazagan** [comment]
```
public void Compose(IContainer container)
    {
        container.PaddingTop(1).Row(row =>
        {
            row.Spacing(4);
            foreach (var item in list)
            {
                row.RelativeItem().Column(column =>
                {
                    foreach (var item2 in list2)
                    {
                        const int Size = 174;
                        const int padding = 5;
                        const int NameSpace = 80;
                        column.Item().PaddingTop(4).Border(1).BorderColor(Colors.Black).Width(Size).Column(column =>
                        {
                            column.Item().PaddingHorizontal(padding).Text(t => TruncatedText(t, Size - padding, 2, item2.Name, 10));
                            column.Item().PaddingHorizontal(padding).PaddingBottom(5).Table(table =>
                            {
                                table.ColumnsDefinition(cd =>
                                {
                                    cd.ConstantColumn(Space);
                                    cd.ConstantColumn(50);
                                    cd.RelativeColumn();
                                });

                                table.Cell().Row(1).Column(1).AlignLeft().Text($"Name").FontSize(9).BackgroundColor(Colors.Grey.Lighten3);
                                table.Cell().Row(1).Column(2).AlignLeft().Text("score").FontSize(9);
                                table.Cell().Row(1).Column(3).AlignRight().Text("Wins").FontSize(9);
                                uint currentRow = 2;
                                foreach (var subItem in item.SubList)
                                {
                                    table.Cell().Row(currentRow).Column(1).BorderBottom(1).BorderColor(Colors.Black).AlignLeft().Text(t => TruncatedText(t, NameSpace, 2, subItem.Name, 9));
                                    table.Cell().Row(currentRow).Column(2).BorderBottom(1).BorderColor(Colors.Black).AlignCenter().Text(subItem.Score.ToString("N0")).FontSize(9);
                                    table.Cell().Row(currentRow).Column(3).BorderBottom(1).BorderColor(Colors.Black).AlignRight().Text($"{subItem.Wins}").FontSize(9);
                                    currentRow++;
                                }
                            });
                        });
                    }
                });
            }
        });
    }
    ```
    Apologies for the look, I have it broken out a lot into sub methods, and this is basically after recombining it, and stripping out some information.

**Comment by Kazagan** [comment]
A follow up question, Is there a way to get the height back from a Container?

**Comment by Kazagan** [comment]
> The `EnsureSpace()` element may help a little. It pushes its child to the next page, if there is less than required vertical space. You may think of it as a less-strict version of the `ShowEntire()` element.

Ensure Space allows it to continue onto the next page, which is something I'm trying to avoid, hopefully in the future by finding a way to split it into a new container with the title appending CONTIUED to the title. I'm trying to find a way to ensure I know when it would need to be split.

**Comment by DawidRobertNowak** [comment, marked as answer]
Hello,

We apologize for the long delay. Since this thread was opened, the library has gone through many releases, and there are now better tools for both of your questions.

For the crash you hit when a box is too tall to fit, PreventPageBreak() is a less-strict alternative to ShowEntire() — it keeps the content together if it fits, and otherwise moves it to the next page rather than throwing:
https://www.questpdf.com/api-reference/prevent-page-break.html

For what you actually described — detecting when content won't fit, splitting it across pages, and repeating the title with a "Continued" suffix — Dynamic Components are the right tool. Their Compose method runs per page with access to the available space, and each element you build exposes its measured Size.Height (this also answers your follow-up about getting a container's height). The "table with per-page subtotals" example shows the full measure-and-split pattern:
https://www.questpdf.com/concepts/code-patterns/dynamic-components.html

We are closing this thread as resolved, but should you run into any further problems, please don't hesitate to open a new issue.

Thank you!
